src/oscar/apps/checkout/mixins.py

```diff
@@ -268,44 +265,15 @@ def send_signal(self, request, response, order):
     def get_success_url(self):
         return reverse('checkout:thank-you')
 
-    def send_confirmation_message(self, order, code, **kwargs):
-        try:
-            ctx = self.get_message_context(order, code)
-        except TypeError:
-            # It seems like the get_message_context method was overridden and
-            # it does not support the code argument yet
-            logger.warning(
-                'The signature of the get_message_context method has changed, '
-                'please update it in your codebase'
-            )
-            ctx = self.get_message_context(order)
-
-        try:
-            event_type = CommunicationEventType.objects.get(code=code)
-        except CommunicationEventType.DoesNotExist:
-            # No event-type in database, attempt to find templates for this
-            # type and render them immediately to get the messages.  Since we
-            # have not CommunicationEventType to link to, we can't create a
-            # CommunicationEvent instance.
-            messages = CommunicationEventType.objects.get_and_render(code, ctx)
-            event_type = None
-        else:
-            messages = event_type.get_messages(ctx)
-
-        if messages and messages['body']:
-            logger.info("Order #%s - sending %s messages", order.number, code)
-            dispatcher = Dispatcher(logger)
-            dispatcher.dispatch_order_messages(order, messages,
-                                               event_type, **kwargs)
-        else:
-            logger.warning("Order #%s - no %s communication event type",
-                           order.number, code)
+    def send_order_placed_email(self, order):
+        extra_context = self.get_message_context(order)
+        dispatcher = Dispatcher(logger=logger)
+        dispatcher.send_order_placed_email_for_user(order, extra_context)
```

Comment:
As noted in my comment on the main thread, I think it is a mistake to move this logic to the Dispatcher. The order app itself needs to be responsible for generating the order placed email.

Reply:
What do you think about the idea suggested by @sasha0 (https://github.com/django-oscar/django-oscar/pull/2591#issuecomment-468375839)?

> What do you think about `CustomerDispatcher` and `OrderDispatcher` for sending notifications, such as `send_order_placed_email_for_user` or `send_password_reset_email_for_user`?

Probably this is the optimal approach to use "special dispatcher" for concrete emails sending - each method that can be easily customised.

Reply:
I think this makes sense - would these classes subclass the core Dispatcher or just use it as an object?

Reply:
I think these classes can use it just as an object.

Reply:
E.g.
```python
from oscar.core.loading import get_class


Dispatcher = get_class('communication.utils', 'Dispatcher')


class CustomerDispatcher:

    # Event codes
    REGISTRATION_EVENT_CODE = 'REGISTRATION'

    dispatcher = Dispatcher()

    def send_registration_email_for_user(self, user, extra_context):
        messages = self.dispatcher.get_messages(
            self.REGISTRATION_EVENT_CODE, extra_context)
        self.dispatcher.dispatch_user_messages(user, messages)
```

Reply:
:+1: